overlord/snapstate/autorefresh_test.go

```diff
@@ -439,6 +449,86 @@ func (s *autoRefreshTestSuite) TestLastRefreshRefreshHoldExpired(c *C) {
 	c.Assert(config.IsNoOption(err), Equals, true)
 }
 
+func (s *autoRefreshTestSuite) TestLastRefreshRefreshHoldExpiredButResetWhileLockUnlocked(c *C) {
+	s.state.Lock()
+	defer s.state.Unlock()
+
+	t0 := time.Now()
+	twelveHoursAgo := t0.Add(-12 * time.Hour)
+	fiveMinutesAgo := t0.Add(-5 * time.Minute)
+	oneHourInFuture := t0.Add(time.Hour)
+	s.state.Set("last-refresh", twelveHoursAgo)
+
+	holdTime := fiveMinutesAgo
+	tr := config.NewTransaction(s.state)
+	tr.Set("core", "refresh.hold", holdTime)
+	tr.Commit()
+
+	logbuf, restore := logger.MockLogger()
+	defer restore()
+
+	sent := false
+	ch := make(chan struct{})
+	// make the store snap action function trigger a background go routine to
+	// change the held-time underneath the auto-refresh
+	go func() {
+		// wait to be triggered by the snap action ops func
+		<-ch
+		s.state.Lock()
+		defer s.state.Unlock()
+
+		// now change the refresh.hold time to be an hour in the future
+		tr := config.NewTransaction(s.state)
+		tr.Set("core", "refresh.hold", oneHourInFuture)
+		tr.Commit()
+
+		// trigger the snap action ops func to proceed
+		ch <- struct{}{}
+	}()
+
+	s.store.snapActionOpsFunc = func() {
+		// only need to send once, this will be invoked multiple times for
+		// multiple snaps
+		if !sent {
+			ch <- struct{}{}
+			sent = true
+			// wait for a response to ensure that we block waiting for the new
+			// refresh time to be committed in time for us to read it after
+			// returning in this go routine
+			<-ch
+		}
+	}
+	defer func() {
+		s.store.snapActionOpsFunc = nil
+	}()
+
+	af := snapstate.NewAutoRefresh(s.state)
+	s.state.Unlock()
+	err := af.Ensure()
+	s.state.Lock()
+	c.Check(err, IsNil)
+
+	var lastRefresh time.Time
+	s.state.Get("last-refresh", &lastRefresh)
+	c.Check(lastRefresh.Year(), Equals, time.Now().Year())
+
+	// hold was reset mid-way to a new value one hour into the future
+	tr = config.NewTransaction(s.state)
+	var t1 time.Time
+	err = tr.Get("core", "refresh.hold", &t1)
+	c.Assert(err, IsNil)
+
+	// when traversing json through the core config transaction, there will be
+	// different wall/monotonic clock times, we remove this ambiguity by
+	// formatting as rfc3339 which will strip this negligible difference in time
+	c.Assert(t1.Format(time.RFC3339), Equals, oneHourInFuture.Format(time.RFC3339))
+
+	// we shouldn't have had a message about "all snaps are up to date", we
+	// should have a message about being aborted mid way
+	c.Assert(strings.Contains(logbuf.String(), "Auto-refresh was delayed mid-way through launching, aborting to try again later"), Equals, true)
```

Comment:
I think testutil.Containts can be used for these checks

Reply:
no this is actually wrong, testutil.Contains just checks for list membership, here we need to check for a pattern, but because the source string has a newline in it, go-check's Matches mucks things up with it's silly newline regex thing it does, so if you don't like strings.Contains then I will need to go fight go-check some more to fix this

if I had a penny for every time I have had to sit and fight go-check's Matches to get the right regex newline incantation I would be have quite a few pennies

Reply:
I gave up and just used regexp directly instead :man_shrugging: :man_shrugging: :man_shrugging: :man_shrugging: :man_shrugging: :man_shrugging: :man_shrugging: :man_shrugging: :man_shrugging: :man_shrugging:

Reply:
mmh, did I read this wrong:

https://github.com/snapcore/snapd/blob/master/testutil/containschecker.go#L71

it looks like it should support also string/string

Reply:
I don't know why it didn't work when I first applied the change but you are correct of course

Reply:
It surely does substring checks, that's my very first contribution to snapd :)